vault/plugin_catalog.go

```diff
@@ -89,7 +89,28 @@ type pluginClient struct {
 
 func wrapFactoryCheckPerms(core *Core, f logical.Factory) logical.Factory {
 	return func(ctx context.Context, conf *logical.BackendConfig) (logical.Backend, error) {
-		if err := core.CheckPluginPerms(conf.Config["plugin_name"]); err != nil {
+		pluginName := conf.Config["plugin_name"]
+		pluginVersion := conf.Config["plugin_version"]
+		pluginTypeRaw := conf.Config["plugin_type"]
+		pluginType, err := consts.ParsePluginType(pluginTypeRaw)
+		if err != nil {
+			return nil, err
+		}
+
+		pluginDescription := fmt.Sprintf("%s plugin %s", pluginTypeRaw, pluginName)
+		if pluginVersion != "" {
+			pluginDescription += " version " + pluginVersion
+		}
+
+		plugin, err := core.pluginCatalog.Get(ctx, pluginName, pluginType, pluginVersion)
+		if err != nil {
+			return nil, fmt.Errorf("failed to find %s in plugin catalog: %w", pluginDescription, err)
+		}
+		if plugin == nil {
+			return nil, fmt.Errorf("failed to find %s in plugin catalog", pluginDescription)
+		}
+
+		if err := core.CheckPluginPerms(path.Base(plugin.Command)); err != nil {
```

Comment:
Is `path.Base()` just an extra safeguard here? I thought the register endpoint rejects any command strings that contain directories.

Reply:
The `Command` field is stored in the plugin entry without the directory, but the `Get` method then adds the plugin directory again before it returns the plugin: https://github.com/hashicorp/vault/blob/main/vault/plugin_catalog.go#L732

Your comment prompted me to check how reading a plugin from the API is handled, because I know we want a plugin entry to be round-trippable for TFVP. The plugin read API handler does also strip off the directory, but it uses a slightly different method to do so:
https://github.com/hashicorp/vault/blob/main/vault/logical_system.go#L565

So I'll update this to use the same method for consistency. It _should_ give the same result, because we disallow subdirectories and evaluate symlinks within the plugin directory, but `filepath.Rel` is more future proof if we ever start using subdirectories in the future.

Reply:
Oh interesting, thanks for the explanation!